added "Actionable" query to `athena.tf` with appropriate Terraform syntax to accommodate 4 workgroups

austinlasseter opened the pull request:
@chrisgilmerproj -- I updated the "policy freqs" query, and I also added 2 others queries to match the examples in https://dds.atlassian.net/wiki/spaces/SEN/pages/901152780/Table+View+scopes.

In my text editor (Atom) all the code is green starting at line 40, which probably indicates there's a missing `}` someplace after the `query = <<-EOT` but I wasn't able to spot what was causing this.
At line 130 I attempted something new: 
```description = "Policy Freqs last 720 hours HHS"```
this is a best-guess. In the Athena console for Saved Queries, it would be helpful to have a description in addition to the query name. please let me know if there's a better way.

chrisgilmerproj commented:
@austinlasseter - update your title for this PR to be more descriptive please :) Also, looks like tests are failing. Make sure you're running the pre-commit hooks before committing.

austinlasseter commented:
Hi @chrisgilmerproj -- I am having trouble with `pre-commit`. This is what I'm seeing: 
```pre-commit install 
zsh: /usr/local/bin/pre-commit: bad interpreter: /usr/local/Cellar/pre-commit/2.10.1/libexec/bin/python3.9: no such file or directory```
and also 
```pre-commit run -a
zsh: /usr/local/bin/pre-commit: bad interpreter: /usr/local/Cellar/pre-commit/2.10.1/libexec/bin/python3.9: no such file or directory```
I am re-reading your notes from last week in order to try and fix this on my own, just wanted you to know where I'm at.

austinlasseter commented:
Also I need to update the queries (again) based on feedback from Kelli, which is captured [on this confluence page.](https://dds.atlassian.net/wiki/spaces/SEN/pages/863502440/Notes+on+Partition-based+filtering) -- I will push a new commit once all these changes have been made.

chrisgilmerproj commented on .pre-commit-config.yaml:
```diff
@@ -5,7 +5,7 @@ repos:
       - id: check-merge-conflict
       - id: check-yaml
-      - id: detect-aws-credentials
+      # - id: detect-aws-credentials
```
Delete this line, don't just comment it out.

austinlasseter replied:
done. I just pushed a new commit with this small change, and for future repos I'll do the same. I added this information to https://github.com/dod-iac/learn-terraform-aws-instance/blob/main/getting-started.md

chrisgilmerproj replied:
Let's move this Getting Started guide over to `deptofdefense/dds-infra` where it would be seen by more folks.

chrisgilmerproj commented on athena.tf:
```diff
@@ -123,27 +123,70 @@ resource "aws_athena_named_query" "num_records" {
+
+# HHS interesting last 720 hours
+resource "aws_athena_named_query" "hhs-interesting-720" {
+  count       = length(var.subscriber_buckets)
```
This doesn't make sense to me. Is HHS looking at more buckets than just the `sub.hhs.secops` bucket? If not then we should only create one named query and only attach it to the HHS workgroup.

chrisgilmerproj commented on athena.tf:
```diff
@@ -123,27 +123,70 @@ resource "aws_athena_named_query" "num_records" {
+  select client_address, policy, datetime from
+  (SELECT
+  client_address, from_unixtime("start_time"*0.000001)  as datetime, policy
+  from shepherd_global_database_sub_hhs_secops_f23sihm4.dns_data
```
This database only exists once and its tied to a particular workgroup. You can also pull the database name and table name from a terraform parameter.

chrisgilmerproj reviewed (changes requested):
You're adding three new named queries. But they only apply to HHS. So instead of adding them across all workgroups, which would result in 12 new resources, just add them to the workgroup they belong to.

On the other hand, if you can generalize the query to be something that each workgroup can use then let's do that instead. What you have now hardcodes queries that are useless in 3 of the 4 workgroups we provision.

austinlasseter commented:
@chrisgilmerproj -- I have updated per your review. Notes:

- I kept it to a single query for now, rather than 3.
- I attempted to imitate the way you used ".format" in  "num_records" so that the same query would appear in 4 workgroups but with the appropriate table each time.
- I added a description.
- I was uncertain how "<<-EOT" would combine with ".format" so I took my best guess. It passed `terraform fmt` so I think it's okay.

austinlasseter commented:
@chrisgilmerproj --
Questions about github branching:
-- Is it correct that one branch == one pull request, and after the PR is merged I can delete the branch?
-- Is this an okay way to name branches, or do you recommend a bette ? `austin-2102251400`

chrisgilmerproj commented on athena.tf:
```diff
@@ -37,16 +37,16 @@ resource "aws_athena_named_query" "create_view" {
+      %{endif~}
+      %{endfor~}
+      GO
+  EOT
```
There isn't really a reason to modify the spacing here unless you want it to appear with spaces in the query when we update it with terraform. When using the `<<-EOT` syntax the `-` should be handling the whitespace stripping for us but only if the last `EOT` is on the same line as the rest of the code. So I suggest you revert this change.

austinlasseter replied:
I returned EOT section to flush left.

chrisgilmerproj commented on athena.tf:
```diff
@@ -123,27 +123,25 @@ resource "aws_athena_named_query" "num_records" {
+  workgroup   = aws_athena_workgroup.shepherd[count.index].id
+  database    = split(":", aws_glue_catalog_database.shepherd[count.index].id)[1]
+  query = format(
+    <<-EOT
```
Again, you have too much spacing here. It may look nice in your editor  but all the spacing ends up in the final query as it gets rendered.

austinlasseter replied:
I returned EOT section to flush left.

chrisgilmerproj commented:
> @chrisgilmerproj --
> Questions about github branching:
> -- Is it correct that one branch == one pull request, and after the PR is merged I can delete the branch?
> -- Is this an okay way to name branches, or do you recommend a bette ? `austin-2102251400`

Let's do questions like these in slack.

- Yes, one branch == one PR.
- I recommend using your name or initials to start a branch name and then anything that is identifiable (like a JIRA ticket or the name of the project). Using date strings will lead to confusion and probably error.
- Once a PR is merged it should auto delete from github in this repo.

austinlasseter commented:
@chrisgilmerproj -- I returned EOT section to flush left. This should be ready to go now.

chrisgilmerproj reviewed (changes requested):
I've updated the query to be run in athena. Previously there was an issue casting the datetime. The queries here now more closely map to what the CSV job is doing.

chrisgilmerproj commented:
As a reference, here is the interesting query being run by the CSV job:

```sql
SELECT *
FROM 
    (SELECT start_time,
         resolver,
         client_address,
         request_length,
         res_serial,
         elapsed_time,
         response_length,
         client_port,
         address_family,
         core_domain,
         client_qname,
         service_view,
         server_address,
         policy,
         policies,
         rtt,
         flags,
         meta_timestamp,
         meta_engine_type,
         meta_engine_version,
         meta_anonymizer_key_hash,
         meta_customer_name,
         meta_node_id,
         dns_opcode,
         dns_flags,
         dns_rcode,
         dns_id,
         dns_authority,
         dns_additional,
         dns_question_rdtype,
         dns_question_name,
         dns_question_rdclass,
         dns_answer,
         dns_answer_ips,
         dns_simple_answer,
         dns_simple_authority,
         tld,
         client_cc,
         client_region,
         client_city,
         client_asnum,
         client_asname,
         rec_type,
         data_version,
         subscriber,
         year,
         month,
         day,
         hour
    FROM "shepherd_global_database_sub_dod_dds_r8cf2j5q"."dns_data"
    CROSS JOIN unnest(parent_policies) AS t (policy)
    WHERE parent_policies is NOT null
            AND hour >= 1611964800)a
WHERE a.policy IN ('sb-infected-page', 'sb-phishing-page', 'sb-safe-search-youtube', 'sb-safe-search', 'sb-restricted-schedule', 'sb- whitelist')
```

chrisgilmerproj reviewed (approved):
🚀
